types/mempool/mempool_test.go

```diff
@@ -0,0 +1,659 @@
+
+	return res, nil
+}
+
+var (
+	_ sdk.Tx                  = (*testTx)(nil)
+	_ mempool.Tx              = (*testTx)(nil)
+	_ signing.SigVerifiableTx = (*testTx)(nil)
+	_ cryptotypes.PubKey      = (*testPubKey)(nil)
+)
+
+func (tx testTx) GetHash() [32]byte { return tx.hash }
+
+func (tx testTx) Size() int64 { return 1 }
+
+func (tx testTx) GetMsgs() []sdk.Msg { return nil }
+
+func (tx testTx) ValidateBasic() error { return nil }
+
+func (tx testTx) String() string {
+	return fmt.Sprintf("tx a: %s, p: %d, n: %d", tx.address, tx.priority, tx.nonce)
+}
+
+type sigErrTx struct {
+	getSigs func() ([]txsigning.SignatureV2, error)
+}
+
+func (_ sigErrTx) Size() int64 { return 0 }
+
+func (_ sigErrTx) GetMsgs() []sdk.Msg { return nil }
+
+func (_ sigErrTx) ValidateBasic() error { return nil }
+
+func (_ sigErrTx) GetSigners() []sdk.AccAddress { return nil }
+
+func (_ sigErrTx) GetPubKeys() ([]cryptotypes.PubKey, error) { return nil, nil }
+
+func (t sigErrTx) GetSignaturesV2() ([]txsigning.SignatureV2, error) { return t.getSigs() }
+
+func TestDefaultMempool(t *testing.T) {
+	ctx := sdk.NewContext(nil, tmproto.Header{}, false, log.NewNopLogger())
+	accounts := simtypes.RandomAccounts(rand.New(rand.NewSource(0)), 10)
+	txCount := 1000
+	var txs []testTx
+
+	for i := 0; i < txCount; i++ {
+		acc := accounts[i%len(accounts)]
+		tx := testTx{
+			address:  acc.Address,
+			priority: rand.Int63(),
+		}
+		txs = append(txs, tx)
+	}
+
+	// same sender-nonce just overwrites a tx
+	mp := mempool.NewPriorityMempool()
+	for _, tx := range txs {
+		ctx.WithPriority(tx.priority)
+		err := mp.Insert(ctx, tx)
+		require.NoError(t, err)
+	}
+	require.Equal(t, len(accounts), mp.CountTx())
+
+	// distinct sender-nonce should not overwrite a tx
+	mp = mempool.NewPriorityMempool()
+	for i, tx := range txs {
+		tx.nonce = uint64(i)
+		err := mp.Insert(ctx, tx)
+		require.NoError(t, err)
+	}
+	require.Equal(t, txCount, mp.CountTx())
+
+	sel, err := mp.Select(nil, 13)
+	require.NoError(t, err)
+	require.Equal(t, 13, len(sel))
+
+	// a tx which does not implement SigVerifiableTx should not be inserted
+	tx := &sigErrTx{getSigs: func() ([]txsigning.SignatureV2, error) {
+		return nil, fmt.Errorf("error")
+	}}
+	require.Error(t, mp.Insert(ctx, tx))
+	require.Error(t, mp.Remove(tx))
+	tx.getSigs = func() ([]txsigning.SignatureV2, error) {
+		return nil, nil
+	}
+	require.Error(t, mp.Insert(ctx, tx))
+	require.Error(t, mp.Remove(tx))
+
+	// removing a tx not in the mempool should error
+	mp = mempool.NewPriorityMempool()
+	require.NoError(t, mp.Insert(ctx, txs[0]))
+	require.ErrorIs(t, mp.Remove(txs[1]), mempool.ErrTxNotFound)
+	mempool.DebugPrintKeys(mp)
+
+	// inserting a tx with a different priority should overwrite the old tx
+	newPriorityTx := testTx{
+		address:  txs[0].address,
+		priority: txs[0].priority + 1,
+		nonce:    txs[0].nonce,
+	}
+	require.NoError(t, mp.Insert(ctx, newPriorityTx))
+	require.Equal(t, 1, mp.CountTx())
+}
+
+type txSpec struct {
+	i int
+	h int
```

Comment:
what's h?

Reply:
A left over artifact from when I was using tx hashes in an earlier implementation, deleting.